### PR TITLE
cassandra warning about netty's native library

### LINKED ISSUE #3428: Cassandra driver error when starting Cassandra microsevice with docker-compose file
##### **Overview of the issue**

When starting a Cassandra microservice with the docker-compose generated files, the following warning and stack trace appears in the log:

```
2016-04-19 14:16:59.249  INFO 7 --- [           main] c.j.test.config.MetricsConfiguration     : Initializing Metrics Log reporting
2016-04-19 14:17:05.551  INFO 7 --- [ost-startStop-1] com.mycompany.myapp.config.WebConfigurer   : Web application configuration, using profiles: [prod]
2016-04-19 14:17:05.579  INFO 7 --- [ost-startStop-1] com.mycompany.myapp.config.WebConfigurer   : Web application fully configured
2016-04-19 14:17:07.081  INFO 7 --- [ost-startStop-1] ccom.mycompany.myapp.CasstestApp            : Running with Spring profile(s) : [prod]
2016-04-19 14:17:08.003  INFO 7 --- [           main] c.j.test.config.LoggingConfiguration     : Initializing Logstash logging
2016-04-19 14:17:09.065  WARN 7 --- [           main] com.datastax.driver.core.NettyUtil       : Found Netty's native epoll transport in the classpath, but epoll is not available. Using NIO instead.
java.lang.UnsatisfiedLinkError: no netty-transport-native-epoll in java.library.path
at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1864)
at java.lang.Runtime.loadLibrary0(Runtime.java:870)
at java.lang.System.loadLibrary(System.java:1122)
at io.netty.util.internal.NativeLibraryLoader.load(NativeLibraryLoader.java:168)
at io.netty.channel.epoll.Native.<clinit>(Native.java:49)
at io.netty.channel.epoll.Epoll.<clinit>(Epoll.java:30)
at java.lang.Class.forName0(Native Method)
at java.lang.Class.forName(Class.java:264)
at com.datastax.driver.core.NettyUtil.<clinit>(NettyUtil.java:68)
at com.datastax.driver.core.NettyOptions.eventLoopGroup(NettyOptions.java:101)
at com.datastax.driver.core.Connection$Factory.<init>(Connection.java:695)
at com.datastax.driver.core.Cluster$Manager.init(Cluster.java:1286)
at com.datastax.driver.core.Cluster.init(Cluster.java:159)
at com.datastax.driver.core.Cluster.connect(Cluster.java:249)
at com.datastax.driver.core.Cluster.connect(Cluster.java:282)
at org.springframework.cassandra.config.CassandraCqlSessionFactoryBean.afterPropertiesSet(CassandraCqlSessionFactoryBean.java:82)
at org.springframework.data.cassandra.config.CassandraSessionFactoryBean.afterPropertiesSet(CassandraSessionFactoryBean.java:43)
at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1637)
at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1574)
at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:545)
at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:482)
at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:306)
at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230)
at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:302)
at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:197)
at org.springframework.beans.factory.support.DefaultListableBeanFactory.findAutowireCandidates(DefaultListableBeanFactory.java:1192)
at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1116)
at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1014)
at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredFieldElement.inject(AutowiredAnnotationBeanPostProcessor.java:545)
at org.springframework.beans.factory.annotation.InjectionMetadata.inject(InjectionMetadata.java:88)
at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor.postProcessPropertyValues(AutowiredAnnotationBeanPostProcessor.java:331)
at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.populateBean(AbstractAutowireCapableBeanFactory.java:1214)
at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:543)
at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:482)
at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:306)
at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230)
at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:302)
at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:197)
at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:772)
at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:839)
at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:538)
at org.springframework.boot.context.embedded.EmbeddedWebApplicationContext.refresh(EmbeddedWebApplicationContext.java:118)
at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:766)
at org.springframework.boot.SpringApplication.createAndRefreshContext(SpringApplication.java:361)
at org.springframework.boot.SpringApplication.run(SpringApplication.java:307)
at com.mycompany.myapp.CasstestApp.main(CasstestApp.java:71)
at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at java.lang.reflect.Method.invoke(Method.java:498)
at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:54)
at java.lang.Thread.run(Thread.java:
```

This does not seems to prevent the app to start, but the application is marked has _down_ in the registry and those can not be reached.

##### **JHipster Version(s)** 
master + fixes from #3420 +  manually workaround of #3427

##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder** 

3 directories: 
- jhigate
- casstest (cassandra microservice)
- compose (docker-compose config)

**gateway:**
```
{
  "generator-jhipster": {
    "jhipsterVersion": "3.0.0",
    "baseName": "jhigate",
    "packageName": "com.mycompany.myapp",
    "packageFolder": "com/mycompany/myapp",
    "serverPort": "8080",
    "authenticationType": "jwt",
    "hibernateCache": "hazelcast",
    "clusteredHttpSession": "no",
    "websocket": "no",
    "databaseType": "sql",
    "devDatabaseType": "h2Disk",
    "prodDatabaseType": "mysql",
    "searchEngine": "no",
    "buildTool": "maven",
    "jwtSecretKey": "080bce90a1e6565afd977f09ce38b41dc315b080",
    "useSass": false,
    "applicationType": "gateway",
    "testFrameworks": [
      "gatling"
    ],
    "jhiPrefix": "jhi",
    "enableTranslation": true,
    "nativeLanguage": "en",
    "languages": [
      "en"
    ]
  }
}
```

**cassandra microservice:**
```
{
  "generator-jhipster": {
    "jhipsterVersion": "3.0.0",
    "baseName": "casstest",
    "packageName": "com.mycompany.myapp",
    "packageFolder": "com/mycompany/myapp",
    "serverPort": "8085",
    "authenticationType": "jwt",
    "hibernateCache": "no",
    "databaseType": "cassandra",
    "devDatabaseType": "cassandra",
    "prodDatabaseType": "cassandra",
    "searchEngine": "no",
    "buildTool": "maven",
    "jwtSecretKey": "a310878c074bd02c67adc51c484c420978c039aa",
    "enableTranslation": false,
    "applicationType": "microservice",
    "testFrameworks": [
      "gatling"
    ],
    "jhiPrefix": "jhi",
    "skipClient": true,
    "skipUserManagement": true,
    "clusteredHttpSession": "no",
    "websocket": "no",
    "enableSocialSignIn": false
  }
}
```

**compose:**
```
{
  "generator-jhipster": {
    "appsFolders": [
      "casstest",
      "jhigate"
    ],
    "directoryPath": "../",
    "useElk": true,
    "jwtSecretKey": "1523fb4eff5ba4c0966a8c69a10bd136e0a9d37d"
  }
}
```

##### **Browsers and Operating System** 
OS/X

##### **Reproduce the error** 

Run yo jhispter in app subdirectory
Run `./mvnw package -Pprod -DskipTests docker:build` in the gateway and casstest directories
Run `yo jhipster:docker-compose` in compose directory
Run `docker-compose up -d` in compose directory to start the cluster and all containers
Apply all cql scripts to the Cassandra cluster
Start the casstest service again: `docker-compose up -d casstest-app `

See the error in the logs.

### PR DESCRIPTION
**Pull request by raphaelbrugier**
fix #3428

### CODE REVIEW

File: generators/server/templates/_build.gradle [hunk 1]
```diff
@@ -102,6 +102,11 @@ description = ''
 
 configurations {
     providedRuntime
+    <%_ if (applicationType == 'microservice' || applicationType == 'gateway' || applicationType === 'uaa') { _%>
+    // netty's native is pulled by spring-cloud-starter-ribbon, but is useless unless you explicitly add the native binary dependency.
+    // Having it in the classpath without the binary can cause warnings
+    all*.exclude group: 'io.netty', module: 'netty-transport-native-epoll'
```
Comment (deepu105) [inline generators/server/templates/_build.gradle]: if it affects only cassandra shouldnt this be done only when DB is cassandra?

Comment (raphaelbrugier) [inline generators/server/templates/_build.gradle]: The warning only shows up with the Cassandra java driver, but this library is not the implementation used by Netty unlike the user explicitly add another lib with the binary.

Like this for linux:
```
    <dependency>
      <groupId>io.netty</groupId>
      <artifactId>netty-transport-native-epoll</artifactId>
      <version>${project.version}</version>
      <classifier>linux-x86_64</classifier>
    </dependency>
```

So I think it's better to remove it from the classpath, even if Cassandra is not present, and let the user decide if he really wants to activate the native support.
In that case he will add the 2 dependencies.

Comment (deepu105) [inline generators/server/templates/_build.gradle]: ok makes sense

**Comment by deepu105** [comment]
LGTM :+1:

**Pull request merged**